src/crypto/CHIPCryptoPAL.h

```diff
@@ -1804,6 +1805,74 @@ CHIP_ERROR ExtractSubjectFromX509Cert(const ByteSpan & certificate, MutableByteS
  **/
 CHIP_ERROR ExtractIssuerFromX509Cert(const ByteSpan & certificate, MutableByteSpan & issuer);
 
+class PemEncoder
+{
+public:
+    static constexpr size_t kNumBytesPerLine   = 48u;
+    static constexpr size_t kMinLineBufferSize = 64u + 1u; // PEM expects 64 characters wide and a null terminator at least.
+    static_assert(kMinLineBufferSize == (BASE64_ENCODED_LEN(kNumBytesPerLine) + 1),
+                  "Internal incoherence of library configuration!");
+
+    /**
+     * @brief Construct a PEM encoder for element type `encodedElement` whose DER data is in `derBytes` span.
+     *
+     * @param encodedElement - Element type string to include in header/footer (e.g. "CERTIFICATE"). Caller must provide correct
+     * uppercase.
+     * @param derBytes - Byte span containing data to encode. May be empty.
+     */
+    explicit PemEncoder(const char * encodedElement, ByteSpan derBytes) : mEncodedElement(encodedElement), mDerBytes(derBytes) {}
+
+    // No copies.
+    PemEncoder(const PemEncoder &)             = delete;
+    PemEncoder & operator=(const PemEncoder &) = delete;
+
+    /**
+     * @brief Returns the next line of the encoding, if anything left to do.
+     *
+     * If `destSize` == 0, nothing will be written. Otherwise if `destSize < kMinLineBufferSize`,
```

Comment:
Could we re-make the API to seem friendlier with the caller? This seems to be biased towards easier implementation describing `kMinLineBufferSize`.

I cannot think of a situation when the `if destSize < kMinLineBufferSize` is desirable. Maybe it would be better to define a `class PemLine : StringBuilder<kMinLineBufferSize>` (or similar really .... it is just a buffer with storage) and use that as argument, so we are always guaranteed a kMinLineBufferSize and not worry about these awkward "what if the buffer size is various sizes". Then the API will be:
   - guaranteed a new line will be read
   - we can avoid an internal buffer if we control the output class
   - callers can use the `c_str` out of it

Reply:
Maybe we could just make an iterator that has a buffer built in. Really I want to write:

```cpp
PemEncoder encoder("CERTIFICATE", TestCerts::sTestCert_PAA_FFF1_Cert);

for (auto line : encoder)
{
   lines_vec.push_back(line.c_str());
}
```

And not deal with the buffering bits.